### ISSUE TITLE
Commands for creating transactions from non-wallet addresses

### ISSUE THREAD

**Issue by dexX7** [OmniLayer/omnicore#61]
> @achamely: once native commands for creating tx's from unknown address is available we'll use it. But at the moment OmniCore needs to have the address added/scanned before it will create tx's. Unfortunately for Omniwallet, this isn't possible. 1: too many addresses, 2: we don't have users private keys to add.

Please help me to understand this a bit more. While I can imagine what you may need, it's probably better to clearly define some requirements.

Right now, we are able to construct and sign any transaction for wallet addresses. Those transactions can be broadcasted, or just signed, but currently, this is limited to wallet keys.

As intermediate step, would it help to have an interface to create raw, unsigned transactions, where the caller provides a set of unspent outputs? Would that be possible for OmniWallet?

**Comment by achamely** [comment]
@dexX7 

Omniwallet is currently creating those tx from scratch itself so we already are getting the utxo info (from obelisk atm. Hopefully soon switching to an addr_index bitcoind)

That being said as an intermediate step that could work. We'll be able to cleanup / remove a lot of the manual tx creation code we are using and start using core as the source

**Comment by zathras-crypto** [comment]
Just to add my 2 cents in here:

* I'm anti including addrindex stuff in OmniCore because I think it bloats the codebase and introduces potential risks and bugs for every potential user to satisfy only a single use-case (web wallets).  Additionally since Sipa (the original author of addrindex) himself didn't want addrindex to be merged into the main code this gives me reservations on scalability etc.
* I do want to help the OmniWallet guys be as successful as possible and that includes giving them the ability to create any transaction via OmniCore for any address, but this needs to be balanced with potential negative impacts to OmniCore 
* I don't buy the argument that addrindex is needed for P2SH, I believe use of watch-only functionality can be used to add a P2SH address and obtain UTXOs for constructing sends
* @CraigSellars suggested the possibility of adding addrindex functionality, but disabling it by default - this seems like it could be a potential compromise.

Thanks
Z

**Comment by dexX7** [comment]
When I created the topic, I primarily had in mind to support transaction creation, where unspent outputs are provided, similar to `createrawtransaction`,

> I'm anti including addrindex stuff in OmniCore

+1 (mostly)

It's just bloat, if those features are not used, and if the only goal is to fetch unspent outputs, for the creation of non-wallet transactions, then an address-/transaction-address-index is over the top, and what's actually needed is an index, which associates addresses with unspent outputs.

So basically three modes are thinkable:

 1. wallet scope
 2. unspent outputs-address index (such as https://github.com/bitcoin/bitcoin/pull/5048)
 3. full blown index with transaction histories for all addresses

It sounds like OmniWallet only needs (2), and not (3), but please correct me, if I'm wrong.

However, I do think it is probably of value to provide indirect support for (2) or (3), maybe in form of more than one release, or by having another branch with those extensions, which is kept conflict-free.

FWIW, I recently rebased the address-index stuff:

- [Bitcoin Core 0.9 with address index](https://github.com/dexX7/bitcoin/tree/addrindex-0.9)
- [Bitcoin Core 0.10 with address index](https://github.com/dexX7/bitcoin/tree/addrindex-0.10)
- [Omni Core 0.0.9 (release) with address index](https://github.com/dexX7/bitcoin/tree/addrindex-mscore-0.0.9)
- [Omni Core 0.0.10 (not up-to-date) with address index](https://github.com/dexX7/bitcoin/tree/addrindex-omnicore-0.0.10)

> I don't buy the argument that addrindex is needed for P2SH

Can you please elaborate? What's the relation between P2SH and the address index? And speaking of, I somewhere read Omni Core should get a better P2SH support. What does this mean? A few details would be great, about what you guys have in mind here.. :)

**Comment by zathras-crypto** [comment]
> Can you please elaborate? What's the relation between P2SH and the address index? And speaking of, I somewhere read Omni Core should get a better P2SH support. What does this mean? A few details would be great, about what you guys have in mind here.. :)

Sure, there is this repo https://github.com/mastercoin-MSC/msc-scripts-n-tricks/blob/master/mastercoin/p2sh which contains python scripts for sending P2SH transactions.  I don't know them very well myself (I'm 'anti' external scripts and 'pro' building those types of functions into Omni Core) but I do know they require to specify a UTXO to use with the send.  My point was that I believe (not 100%) that by adding the sending P2SH address to watch-only we can thus avoid the requirement to manually specify a UTXO and allow P2SH transactions to be created directly from OmniCore without requiring the keys in the wallet or an external script.

To be honest though, I've been focusing on so many other things sending P2SH (and those scripts) is something I'm not very familiar with yet.

**Comment by dexX7** [comment]
Ah, thanks!

The commands `send_MP`, `sendrawtx_MP` should actually just work fine, and a P2SH address can simply be used instead of a regular one. In this context some old tutorial: [Using Master Core to create a P2SH multisig wallet and send some tokens](https://gist.github.com/dexX7/96c7ca99f909c15fad68) (sending BTC as raw transaction is done for fun, but not required)

In this case it looks like raw Omni transactions are created, which are then embedded manually in raw Bitcoin transactions. This doesn't seem to have much to do with P2SH in particular, but (manual) raw transaction construction in general.

**Comment by achamely** [comment]
@dexx yea 2 and 3 are the big ones atm but with regards to p2sh if we already have utxo info coming from a different source and the createrawtx supports the use of the p2sh address then that should work a bit better than the current setup (old python scripts)

**Comment by dexX7** [comment]
> createrawtx supports the use of the p2sh address

All `send` commands do. :)

But it seems essential to be able to explicitly provide unspent outputs, which are not part of the wallet to provide a significant benefit for Omni Wallet. The nice part: once this works, it's also possible to run Omni Core without blockchain, and simply use it to construct transactions (e.g. as replacement for the old scripts, or in an offline environment).

I suggest to tackle the question, whether an address-index, or something else, might be integrated into Omni Core, at some point later, once we are able to deal with non-wallet unspent outputs in general. Adding an index then would mean: providing unspent outputs from an internal, and not external source.

**Comment by zathras-crypto** [comment]
> I suggest to tackle the question, whether an address-index, or something else, might be integrated into Omni Core, at some point later, once we are able to deal with non-wallet unspent outputs in general.

I'd like to investigate watch-only for this as I believe that *may* gives us the "get UTXOs for a non-wallet address" solution using primarily existing code.

Note this is *not* a solution for OmniWallet, because adding every address for every user to a wallet as watch-only likely isn't feasible.  But for most other use cases (eg the most common being "I want to create a transaction for address X without the key in the wallet), I believe adding said address as watch only can expose those UTXOs for us.  Further investigation needed of course.

**Comment by dexX7** [comment]
Yes, I think this should work. It's worth to note that adding new watch-only entries requires rescanning though.

We should probably define what we want to accomplish, given that "using watch-only to retrieve UTXO" provides a solution for an aspect, which wasn't considered until now. I guess this goes sort of hand in hand with raw transactions in general, given that the transaction still needs to be signed (and constructed).

If retrieving UTXO is very high priority, and we don't want to include something more sophisticated, then we could also just go through all entries. That's inefficient, but probably faster than rescanning.

**Comment by zathras-crypto** [comment]
> It's worth to note that adding new watch-only entries requires rescanning though.

That's correct indeed - however also to note is you can 'batch' this if you have multiple (ie add a bunch of addresses, rescan once).

> given that the transaction still needs to be signed (and constructed)

I'm making an assumption here that signing occurs elsewhere (since we're talking about scenarios where the key is not in the wallet) - but yep agreed - I see it as a model for constructing transactions where you don't want the key stored in the wallet, but still want Omni Core to be able to construct your transactions.

> If retrieving UTXO is very high priority, and we don't want to include something more sophisticated, then we could also just go through all entries. That's inefficient, but probably faster than rescanning.

Yep, I've had this on my todo list for many moons - that being to answer the question "how long does it actually take to just iterate the UTXO set for a given unspent output?"

> Edit: here some very rudimentary, old code where I tested direct UTXO access. In this form, it shouldn't be used on mainnet, since the command simply dumps all entries.

Aww nice I'll take a look :)

**Comment by zathras-crypto** [comment]
Have you ever timed ```listalltxouts``` by any chance?

**Comment by dexX7** [comment]
> Have you ever timed listalltxouts by any chance?

I only tested it in regtest mode. Would be interesting to know though. I'll get some data within the next days, but probably not before Friday (I have no fully synced main chain at the very moment).

**Comment by dexX7** [comment]
The time for `gettxoutsetinfo` might serve as estimation, and it roughly takes about.. I don't know, 10-25 seconds? During that call all entries are read from the DB and deserialized. Using a filter on top ("get unspent outputs only for address x") has probably almost no additional performance cost.

**Comment by zathras-crypto** [comment]
> The time for gettxoutsetinfo might serve as estimation, and it roughly takes about.. I don't know, 10-25 seconds? 

That's not too bad at all - if we serve that capability with the caveat that it will take an extended amount of time to return the response, perhaps that's an angle we can use for allowing creation of transactions for any address.

> During that call all entries are read from the DB and deserialized

I had thought the UTXO set was cached in RAM?

**Comment by dexX7** [comment]
> I had thought the UTXO set was cached in RAM?

I think the LevelDB cache is held in memory, and if I'm not mistaken, then with the default cache size of 100 MB, the whole UTXO DB is cached.

**Comment by zathras-crypto** [comment]
Ah that makes sense, thanks...

**Comment by dexX7** [comment]
I started with the work on supporting the creation of raw transactions, and so far it looks like it works. However, I'm wondering about the following points:

- Class B transactions require a public key for redemption, while class C transactions don't. We decide whether to build B or C transactions on the fly, based on the user settings. Right now the public key can be provided via RPC optionally. I don't like it, but this seems like the only way.

- What about transactions fees? Should they be completely user defined, or should there be an option, to let Omni Core calculate sufficient fees?

- Does the raw transaction interface need the `referenceamount`, which can be used to send a given amount of BTC to the receiver/reference address?

- We have already `autocommit`, which can be used to disable broadcasting of transactions, and it allows create signed raw transactions. Though it's not based on unspent outputs, but on "addresses", and limited to the wallet. I'm wondering, if there are actually use-cases for this?

- Should the raw transaction interface check the user input, e.g. confirm the sender has sufficient balance, the properties exist, etc., as done for the "create and broadcast" calls?

@achamely: how do you imagine the raw transaction interface should look like? Which values do you want to provide, and which values do you want to retrieve in return?

Let's play it through with a command for simple sends:

```bash
$ ./src/bitcoin-cli "help" "createsend_OMNI"
createsend_OMNI [{"txid":"id","vout":n},...] "toaddress" propertyid "amount" ( fee pubkey )

Creates a raw simple send transaction.

Note: transactions with a payload size up to "datacarriersize" are created as class C transactions, while transactions with larger payloads are created as class B transactions. Providing a public key to redeem multisig dust is only mandatory for class B transactions.

Arguments:
1. outpoints            (string, required) A json array of json objects
     [
       {
         "txid":"hash",   (string, required) the transaction hash
         "vout":n         (number, required) the output number
       }
       ,...
     ]
2. toaddress              (string, required) the address to send to
3. propertyid             (number, required) the identifier of the property to transfer
4. amount                 (string, required) the amount to transfer
5. fee                    (number, optional) the transaction fee (default: 0.0001 BTC)
6. pubkey                 (string, optional) a public key that can spent multisig dust

Result:
transaction               (string) the hex-encoded raw transaction
```
```bash
$ ./src/bitcoin-cli "createsend_OMNI" '[{"txid":"e4de79bd3af00c9ee3e48ef68fc9ba9923916cc839917239353a7101770d70c4","vout":0}]' "mifjkPTqoLEmttH3vcB7GssuRRwPu8T7Dn" 1 "15.00" 0.0005
0100000001c4700d7701713a3539729139c86c912399bac98ff68ee4e39e0cf03abd79dee40000000000ffffffff030000000000000000146a126f6d00000000000000010000000059682f0022020000000000001976a9142291a5ee145781e3572f54c346097de89e939ed588ac0e2bfa02000000001976a9140f931dae288457388bb0604a340d58f793f2dd0c88ac00000000
```
```bash
$ ./src/bitcoin-cli "signrawtransaction" "0100000001c4700d7701713a3539729139c86c912399bac98ff68ee4e39e0cf03abd79dee40000000000ffffffff030000000000000000146a126f6d00000000000000010000000059682f0022020000000000001976a9142291a5ee145781e3572f54c346097de89e939ed588ac0e2bfa02000000001976a9140f931dae288457388bb0604a340d58f793f2dd0c88ac00000000"
{
    "hex" : "0100000001c4700d7701713a3539729139c86c912399bac98ff68ee4e39e0cf03abd79dee4000000006a47304402203ca22913408fdd4fb0ec2e7eb709d56d8b29212effcb462443ea8c9b6f8568ac0220774b115162aa84fa2d84162709a68b11a082d7c95053e1aa08762b73484a781301210398128b2aa3c758a251b81f5a0eba19b94f49d1271876d1cd42b9d6fd6161481dffffffff030000000000000000146a126f6d00000000000000010000000059682f0022020000000000001976a9142291a5ee145781e3572f54c346097de89e939ed588ac0e2bfa02000000001976a9140f931dae288457388bb0604a340d58f793f2dd0c88ac00000000",
    "complete" : true
}
```
```bash
$ ./src/bitcoin-cli "sendrawtransaction" "0100000001c4700d7701713a3539729139c86c912399bac98ff68ee4e39e0cf03abd79dee4000000006a47304402203ca22913408fdd4fb0ec2e7eb709d56d8b29212effcb462443ea8c9b6f8568ac0220774b115162aa84fa2d84162709a68b11a082d7c95053e1aa08762b73484a781301210398128b2aa3c758a251b81f5a0eba19b94f49d1271876d1cd42b9d6fd6161481dffffffff030000000000000000146a126f6d00000000000000010000000059682f0022020000000000001976a9142291a5ee145781e3572f54c346097de89e939ed588ac0e2bfa02000000001976a9140f931dae288457388bb0604a340d58f793f2dd0c88ac00000000"
22f4b954d731c46a4984fb4c70f937e34a89bffe970fadc0cc004ff0bd00156f
```
```bash
$ ./src/bitcoin-cli "setgenerate" true 1
[
    "2af06575e8d60d6f739ac132f460f0b32c0735fa42cb7a26a85e0f5a1eb64afe"
]
```
```bash
$ ./src/bitcoin-cli "gettransaction_MP" "22f4b954d731c46a4984fb4c70f937e34a89bffe970fadc0cc004ff0bd00156f"
{
    "txid" : "22f4b954d731c46a4984fb4c70f937e34a89bffe970fadc0cc004ff0bd00156f",
    "sendingaddress" : "mgwJfj4r5FdcXm83qAuAh37Camvhg3nv4j",
    "referenceaddress" : "mifjkPTqoLEmttH3vcB7GssuRRwPu8T7Dn",
    "ismine" : true,
    "confirmations" : 1,
    "fee" : 0.00050000,
    "blocktime" : 1433771402,
    "version" : 0,
    "type_int" : 0,
    "type" : "Simple Send",
    "propertyid" : 1,
    "divisible" : true,
    "amount" : "15.00000000",
    "valid" : true
}
```

**Comment by zathras-crypto** [comment]
> I started with the work on supporting the creation of raw transactions, and so far it looks like it works. 

Nice nice!

> Class B transactions require a public key for redemption, while class C transactions don't. We decide whether to build B or C transactions on the fly, based on the user settings. Right now the public key can be provided via RPC optionally. I don't like it, but this seems like the only way.

I can't really see any alternative :(

> What about transactions fees? Should they be completely user defined, or should there be an option, to let Omni Core calculate sufficient fees?

I think we should allow Omni Core to calculate fees based on existing fee code in the Bitcoin stack - what does the Bitcoin raw transactions API do in this case?

> Does the raw transaction interface need the "referenceamount", which can be used to send a given amount of BTC to the receiver/reference address?

The feature is used to provide the recipient with enough funds to then 'on-send' the tokens elsewhere - if it's not too obstructive it would make sense to include it as an optional parameter as it provides extra functionality.

> We have already autocommit, which can be used to disable broadcasting of transactions, and it allows create signed raw transactions. Though it's not based on unspent outputs, but on "addresses", and limited to the wallet. I'm wondering, if there are actually use-cases for this?

Autocommit is primarily for testing purposes, so we can spit out transactions without actually writing them to the wallet or broadcasting them to the network.  As to it's merits for testing, I guess you guys would need to tell me if it's useful :)

> Should the raw transaction interface check the user input, e.g. confirm the sender has sufficient balance, the properties exist, etc., as done for the "create and broadcast" calls?

In my opinion yep.  I don't think we should support creating transactions that are knowingly invalid.  

Thanks again dude :)

**Comment by dexX7** [comment]
Yesterday I spent some time of manually going through the steps for #11. Instead of starting with "move to output" transaction types, I simulated it with an oracle server, which prevents spending "other outputs of an address". I'll post more details in #11 at some point, so stay tuned. ;)

I learned a few things, which are relevant for this issue/context:

 - [x] Our P2SH support is solid. It's not very user friendly (well, CLI hacking..), but I was able to create a m-of-n address, where more than one nodes have one key each. I was able to construct, and partially sign transactions one machine, and provide the second signature another one. It's also possible to do this completely offline (but see the next point), or on a machine where no keys were previously added to Omni Core.

 - [ ] Simply providing outpoints (hash, vout n) isn't sufficient to create raw transactions in an offline environment, or where the inputs are not yet broadcasted. Providing the scriptPubKey, and input values is required for those cases.

 - [ ] Providing a reference amount is mandatory for some applications, and as you mentioned, it's useful in general.

> I can't really see any alternative :(

Yeah, I guess we have to live with it. Though I think we may add something like `"createrawclassb"`, `"createrawclassc"` for "power users".

> I think we should allow Omni Core to calculate fees based on existing fee code in the Bitcoin stack - what does the Bitcoin raw transactions API do in this case?

It's up to the creator how to deal with it, and the raw transaction interface has no mechanism to calculate fees (besides `"estimatefee"`). I think we may adopt some of the wallet/fee code though, to make it easier for the user. This is meant in the sense that an user shouldn't have to deal with the fees, in most cases.

> In my opinion yep. I don't think we should support creating transactions that are knowingly invalid.

So the issue here is: we can only verify the information, which is available. Consider the case where one would like to create a transaction offline, with an unsynchronized client. That's basically what I was thinking about in this context.

**Comment by achamely** [comment]
@dexX7 (sorry for the delay in responding)

The raw transaction interface would take in the 
from address (and /or the utxo's), i'm guessing you'd extract from address from utxo info
to address
propertyid 
amount 
fee (optional)
pubkey (optional if not defined use sending address)

Error checking in my opinion is a good idea. We'll be doing checks on front end but a secondary check on tx creation is a good flow. 

omni would expect either a fully prepared unsigned tx or an error dictating why the creation failed.  
We then hand the tx off to the users session to sign and once we receive it back broadcast it out.

**Comment by dexX7** [comment]
A few days ago I funded the activation address with 0.0001 BTC, and then I continued the work on the raw transaction API. I'm now able to create raw transactions with #77! :)

It's a bit laborious, and the following isn't intended as guide or so (will probably follow at some point though), but more to write the steps down:

Funding:
```bash
# fund the activation address with 0.1 TOMNI
omni_send "1Pa6zyqnhL6LDJtrkCMi9XmEDNHJ23ffEr" "3Fc5gWzEQh1YGeqVXH6E4GDEGgbZJREJQ3" 2 "0.1"
1137e11996ea3fe690394bc02cef422d10310b6042c956ca7eb95f24277b6fea
```

Creation:
```bash
# prepare the spending transaction, add the 0.0001 BTC input to a new raw transaction
createrawtransaction '[{"txid":"a207a744e12faa66638b63055f7ebcdef01f1036c421a19a93ab9056defa2ef7","vout":0}]' "{}"
0100000001f72efade5690ab939aa121c436101ff0debc7e5f05638b6366aa2fe144a707a20000000000ffffffff0000000000

# add the payload (simple send 0.1 TOMNI)
omni_createrawtx_multisig "0100000001f72efade5690ab939aa121c436101ff0debc7e5f05638b6366aa2fe144a707a20000000000ffffffff0000000000" "00000000000000020000000000989680" "3Fc5gWzEQh1YGeqVXH6E4GDEGgbZJREJQ3" "0394abb6fe6d492495f855ab3fa29acd6be04f347d17597985da0e8c70a26036f2"
0100000001f72efade5690ab939aa121c436101ff0debc7e5f05638b6366aa2fe144a707a20000000000ffffffff02ac020000000000004751210394abb6fe6d492495f855ab3fa29acd6be04f347d17597985da0e8c70a26036f2210258785fd620c3f80f730dc719245d71109e546d1f00f0acf95b338c6109855a1252ae22020000000000001976a914946cb2e08075bcbaf157e47bcb67eb2b2339d24288ac00000000
 
# add a reference/receiver (1Pa6zyqnhL6LDJtrkCMi9XmEDNHJ23ffEr)
omni_createrawtx_reference "0100000001f72efade5690ab939aa121c436101ff0debc7e5f05638b6366aa2fe144a707a20000000000ffffffff02ac020000000000004751210394abb6fe6d492495f855ab3fa29acd6be04f347d17597985da0e8c70a26036f2210258785fd620c3f80f730dc719245d71109e546d1f00f0acf95b338c6109855a1252ae22020000000000001976a914946cb2e08075bcbaf157e47bcb67eb2b2339d24288ac00000000" "1Pa6zyqnhL6LDJtrkCMi9XmEDNHJ23ffEr"
0100000001f72efade5690ab939aa121c436101ff0debc7e5f05638b6366aa2fe144a707a20000000000ffffffff03ac020000000000004751210394abb6fe6d492495f855ab3fa29acd6be04f347d17597985da0e8c70a26036f2210258785fd620c3f80f730dc719245d71109e546d1f00f0acf95b338c6109855a1252ae22020000000000001976a914946cb2e08075bcbaf157e47bcb67eb2b2339d24288ac22020000000000001976a914f7958092f4af2ca5f90ef592f9eb9732d06c814188ac00000000

# add change with 0.0001 BTC fee (it's a no-op in this case)
omni_createrawtx_change "0100000001f72efade5690ab939aa121c436101ff0debc7e5f05638b6366aa2fe144a707a20000000000ffffffff03ac020000000000004751210394abb6fe6d492495f855ab3fa29acd6be04f347d17597985da0e8c70a26036f2210258785fd620c3f80f730dc719245d71109e546d1f00f0acf95b338c6109855a1252ae22020000000000001976a914946cb2e08075bcbaf157e47bcb67eb2b2339d24288ac22020000000000001976a914f7958092f4af2ca5f90ef592f9eb9732d06c814188ac00000000" '[{"txid":"a207a744e12faa66638b63055f7ebcdef01f1036c421a19a93ab9056defa2ef7","vout":0,"scriptPubKey":"a91498a2d17e08ac677dc220b92e0b79406f2f441c2487","value":0.0001}]' "3Fc5gWzEQh1YGeqVXH6E4GDEGgbZJREJQ3" 0.0001
0100000001f72efade5690ab939aa121c436101ff0debc7e5f05638b6366aa2fe144a707a20000000000ffffffff03ac020000000000004751210394abb6fe6d492495f855ab3fa29acd6be04f347d17597985da0e8c70a26036f2210258785fd620c3f80f730dc719245d71109e546d1f00f0acf95b338c6109855a1252ae22020000000000001976a914946cb2e08075bcbaf157e47bcb67eb2b2339d24288ac22020000000000001976a914f7958092f4af2ca5f90ef592f9eb9732d06c814188ac00000000

# sign the transaction with my key
signrawtransaction "0100000001f72efade5690ab939aa121c436101ff0debc7e5f05638b6366aa2fe144a707a20000000000ffffffff03ac020000000000004751210394abb6fe6d492495f855ab3fa29acd6be04f347d17597985da0e8c70a26036f2210258785fd620c3f80f730dc719245d71109e546d1f00f0acf95b338c6109855a1252ae22020000000000001976a914946cb2e08075bcbaf157e47bcb67eb2b2339d24288ac22020000000000001976a914f7958092f4af2ca5f90ef592f9eb9732d06c814188ac00000000" '[{"txid":"a207a744e12faa66638b63055f7ebcdef01f1036c421a19a93ab9056defa2ef7","vout":0,"scriptPubKey":"a91498a2d17e08ac677dc220b92e0b79406f2f441c2487","redeemScript":"542102d797b8526701a3dfdb52d11f89377e0288b14e29b1414d64de065cd337069c3b21036a4caa95ec1d55f1b75a8b6c7345f22b4efc9e25d38ab058ef7d6f60b3b744f7410499e86235a6a98fc295d7cfe641d37409f2840ad32e0211579cae488bd86cf01daf7ad8f082d968ea4bca77e794fffd1a31583f36f37b1198e51d0651cdbcf3214104b7a3d7f7ccdf211dfd180815b87332b4773cc40bff72a4d0bb60f3a85409d19f99709331c6b11c976fe274a86d789a1cf2b3b0be29fe5fc55c93ad9e08459c4f4104e65b098558d637cfcf3194214637f8838338b141259b698d2a027b069d405b6502ad4a4e9aa75094fa431a6c9af580f5917834a6d4cec946054df33194b2967855ae"}]' '["XXX"]'
0100000001f72efade5690ab939aa121c436101ff0debc7e5f05638b6366aa2fe144a707a200000000fd59010047304402206016fcbea57a75a2c6fd3853a52acdf25b5ba6944c6df54119dd5b258d3133cf022029189e5062332640e4114451f40f6845e24c7a412445c84dd6ab90263a1a85c0014d0d01542102d797b8526701a3dfdb52d11f89377e0288b14e29b1414d64de065cd337069c3b21036a4caa95ec1d55f1b75a8b6c7345f22b4efc9e25d38ab058ef7d6f60b3b744f7410499e86235a6a98fc295d7cfe641d37409f2840ad32e0211579cae488bd86cf01daf7ad8f082d968ea4bca77e794fffd1a31583f36f37b1198e51d0651cdbcf3214104b7a3d7f7ccdf211dfd180815b87332b4773cc40bff72a4d0bb60f3a85409d19f99709331c6b11c976fe274a86d789a1cf2b3b0be29fe5fc55c93ad9e08459c4f4104e65b098558d637cfcf3194214637f8838338b141259b698d2a027b069d405b6502ad4a4e9aa75094fa431a6c9af580f5917834a6d4cec946054df33194b2967855aeffffffff03ac020000000000004751210394abb6fe6d492495f855ab3fa29acd6be04f347d17597985da0e8c70a26036f2210258785fd620c3f80f730dc719245d71109e546d1f00f0acf95b338c6109855a1252ae22020000000000001976a914946cb2e08075bcbaf157e47bcb67eb2b2339d24288ac22020000000000001976a914f7958092f4af2ca5f90ef592f9eb9732d06c814188ac00000000
```

Validation:
```bash
# test decoding of the raw transaction
omni_decodetransaction "0100000001f72efade5690ab939aa121c436101ff0debc7e5f05638b6366aa2fe144a707a200000000fd59010047304402206016fcbea57a75a2c6fd3853a52acdf25b5ba6944c6df54119dd5b258d3133cf022029189e5062332640e4114451f40f6845e24c7a412445c84dd6ab90263a1a85c0014d0d01542102d797b8526701a3dfdb52d11f89377e0288b14e29b1414d64de065cd337069c3b21036a4caa95ec1d55f1b75a8b6c7345f22b4efc9e25d38ab058ef7d6f60b3b744f7410499e86235a6a98fc295d7cfe641d37409f2840ad32e0211579cae488bd86cf01daf7ad8f082d968ea4bca77e794fffd1a31583f36f37b1198e51d0651cdbcf3214104b7a3d7f7ccdf211dfd180815b87332b4773cc40bff72a4d0bb60f3a85409d19f99709331c6b11c976fe274a86d789a1cf2b3b0be29fe5fc55c93ad9e08459c4f4104e65b098558d637cfcf3194214637f8838338b141259b698d2a027b069d405b6502ad4a4e9aa75094fa431a6c9af580f5917834a6d4cec946054df33194b2967855aeffffffff03ac020000000000004751210394abb6fe6d492495f855ab3fa29acd6be04f347d17597985da0e8c70a26036f2210258785fd620c3f80f730dc719245d71109e546d1f00f0acf95b338c6109855a1252ae22020000000000001976a914946cb2e08075bcbaf157e47bcb67eb2b2339d24288ac22020000000000001976a914f7958092f4af2ca5f90ef592f9eb9732d06c814188ac00000000" '[{"txid":"a207a744e12faa66638b63055f7ebcdef01f1036c421a19a93ab9056defa2ef7","vout":0,"scriptPubKey":"a91498a2d17e08ac677dc220b92e0b79406f2f441c2487","value":0.0001}]'
{
  "txid": "06bfdf0e98ea9101d423bb7e595a315260914404aea0a43c4ebc0c17cc5c2ecd",
  "fee": "0.00008224",
  "sendingaddress": "3Fc5gWzEQh1YGeqVXH6E4GDEGgbZJREJQ3",
  "referenceaddress": "1Pa6zyqnhL6LDJtrkCMi9XmEDNHJ23ffEr",
  "ismine": true,
  "version": 0,
  "type_int": 0,
  "type": "Simple Send",
  "propertyid": 2,
  "divisible": true,
  "amount": "0.10000000",
  "confirmations": 0
}
```

Does this make sense and sounds like a reasonable flow? @zathras-crypto?

**Comment by zathras-crypto** [comment]
Re the process, if you add the change after you add the reference, don't we run the risk of incorrect reference identification if the change address is not = the sender? (change != sender provides multiple reference candidates & we use highest vout in this case).

Also I notice the first key in the multisig is the recipient rather than the sender - is this intentional?

I love the progress!!!

**Comment by dexX7** [comment]
Thanks!

You have a point re: change. But I'm not sure, how to solve this. Currently it "subtracts the outgoing amounts from the incoming amounts", and this is added as change (if not dust) right away. If I'd add the change first, and then another output, then the change wouldn't be based on the actual outgoing amount anymore. Or I could actually calculate the change, when "building" the transaction to cover the whole transaction nevertheless.

> Also I notice the first key in the multisig is the recipient rather than the sender - is this intentional?

Yes, it was intentional. It's a scripthash sender, so I had to provide a public key.

**Comment by dexX7** [comment]
Optionally a [position for the change](https://github.com/dexX7/bitcoin/blob/1c0160af6d7f2840b7d087cadbae0571a22b5134/src/omnicore/rpcrawtx.cpp#L215-L254) can now be provided. It feels a bit like a hack, but in my opinion it's much better than always adding the change to the end.

**Comment by zathras-crypto** [comment]
Interesting...  

I also notice you require the inputs in ```omni_createrawtx_change``` - given that it is assumed the inputs were already added, and you already obtain the raw tx so far, are we not getting the inputs twice just in different formats (once within the raw tx hex to extend, and then a second time via a param)?  

> but on the RPC level, "omni_createrawtx_change" is an "isolated" call, without much context. The alternative could be that the user explicitly provides the incoming and outgoing amouts, but this kinda defeats the purpose (and he or she may just add an output).

Why does the user need to provide the incoming/outgoing amounts?  Isn't that data already there in the rawtxhex?  Perhaps you could 'auto' it by default (pulling inputs and outputs out of raw tx hex):

* ```omni_createrawtx_change [rawtxhex]``` = use defaults; return change to sender, use some default fee
* ```omni_createrawtx_change [rawtxhex] [changerecipient]``` = send change to custom address, use default fee
* ```omni_createrawtx_change [rawtxhex] [changerecipient] [fee]``` = send change to custom address, use custom fee

>  edit: I'm wondering: maybe it would be better to insert change at the beginning as default?

I like this - simple and effective.

Just thinking out loud here bud, this is awesome progress I'm just trying to think of any ways to streamline it further :)

**Comment by dexX7** [comment]
The example calls you mentioned would be very nice, and thanks for the input!

> Isn't that data already there in the rawtxhex?

Unfortunally not, for two reasons: the transaction may be incomplete (consider wanting to add the change first, and then other outputs - not possible), and the inputs that can be extracted from the raw hex are just outpoints (i.e. txid + output index), which may not be in the chain, so their value is unknown.

> use defaults; return change to sender

That's definitely something to add later, but currently also a bit tricky: the sender selection isn't isolated yet.

> I like this - simple and effective.

Then let's go with it for now! :)

**Comment by dexX7** [comment]
Resolved via #77.
